src/cluster.c

```diff
@@ -852,176 +689,306 @@ static dict *parse_cluster_slots(valkeyClusterContext *cc, valkeyReply *reply) {
-                goto error;
+            /* Move all replicas stored in the first parsed replica. */
+            valkeyClusterNode *replica = dictGetEntryVal(der);
+            if (replica->slaves != NULL) {
+                while (listLength(replica->slaves) > 0) {
+                    listNode *node = listFirst(replica->slaves);
+                    if (listAddNodeTail(primary->slaves, node->value) == NULL) {
+                        return VALKEY_ERR;
+                    }
+                    /* Delete element without freeing the moved cluster node. */
+                    replica->slaves->free = NULL;
+                    listDelNode(replica->slaves, node);
+                    replica->slaves->free = listClusterNodeDestructor;
+                }
+                listRelease(replica->slaves);
+                replica->slaves = NULL;
+            }
+            /* Move replica that was parsed first. */
+            if (listAddNodeHead(primary->slaves, replica) == NULL) {
+                return VALKEY_ERR;
             }
+            /* All replicas for this primary moved, set dict value
+             * to NULL avoiding freeing the moved memory. */
+            dictSetHashVal(replicas, der, NULL);
+        }
+    }
+    return VALKEY_OK;
+}
 
-            // if the address string starts with ":0", skip this node.
-            if (sdslen(part[1]) >= 2 && memcmp(part[1], ":0", 2) == 0) {
-                sdsfreesplitres(part, count_part);
-                count_part = 0;
-                part = NULL;
+/* Parse a node from a single CLUSTER NODES line. Only parse primary nodes if
+ * the 'replica_master_id' argument is NULL, otherwise replicas are parsed and
+ * its master_id is given via 'replica_master_id'. */
+static int parse_cluster_nodes_line(valkeyClusterContext *cc, char *line,
+                                    valkeyClusterNode **parsed_node, char **replica_master_id) {
+    char *p, *id = NULL, *addr = NULL, *flags = NULL, *master_id = NULL,
+             *link_state = NULL, *slots = NULL;
+    // clang-format off
+    /* Find required fields. */
+    int i = 0;
+    while ((p = strchr(line, ' ')) != NULL) {
+        *p = '\0';
+        switch(i++){
+            case 0: id = line; break;
+            case 1: addr = line; break;
+            case 2: flags = line; break;
+            case 3: master_id = line; break;
+            case 7: link_state = line; break;
+        }
+        line = p + 1; /* Start of next field. */
+        if (i == 8) { slots = line; break; }
+    }
+    if (i == 7 && line[0] != '\0') link_state = line;
+    // clang-format on
+
+    if (link_state == NULL) {
+        valkeyClusterSetError(cc, VALKEY_ERR_OTHER, "Mandatory fields missing");
+        return VALKEY_ERR;
+    }
+
+    /* Parse flags. */
+    uint8_t role = VALKEY_ROLE_NULL;
+    while (*flags != '\0') {
+        if ((p = strchr(flags, ',')) != NULL)
+            *p = '\0';
+        if (memcmp(flags, "master", 6) == 0) {
```

Comment:
Can you add an example line in a comment, maybe with the names of the different parts underneath? (flags, role, etc.) It would make it easier to read code like this.

```
a8b79a8b798ab master 192.0.0.2:123@456,example.com,foo=bar ...
node-id       role   ip        port cport hostname flags   ...
```

Reply:
Added comment about which fields are parsed/expected